### PR TITLE
Start comfyui in subprocess instead of multiprocessing

### LINKED ISSUE #100: [BUG] `freeze_support()` error when trying to start comfyui process
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
  File "/Users/dark/ai_projects/stable-diffusion-webui/extensions/sd-webui-comfyui/lib_comfyui/comfyui_pre_main.py", line 5, in <module>
    from lib_comfyui import (
  File "/Users/dark/ai_projects/stable-diffusion-webui/extensions/sd-webui-comfyui/lib_comfyui/argv_conversion.py", line 2, in <module>
    from lib_comfyui import webui_settings
  File "/Users/dark/ai_projects/stable-diffusion-webui/extensions/sd-webui-comfyui/lib_comfyui/webui_settings.py", line 3, in <module>
    from lib_comfyui import ipc
  File "/Users/dark/ai_projects/stable-diffusion-webui/extensions/sd-webui-comfyui/lib_comfyui/ipc.py", line 38, in <module>
    'webui': parallel_utils.CallbackWatcher(parallel_utils.CallbackQueue(call_fully_qualified)),
  File "/Users/dark/ai_projects/stable-diffusion-webui/extensions/sd-webui-comfyui/lib_comfyui/parallel_utils.py", line 45, in __init__
    manager = ctx.Manager()
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/managers.py", line 562, in start
    self._process.start()
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/dark/dev_lib/miniconda3/envs/sd/lib/python3.10/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

### PR DESCRIPTION
**Pull request by ljleb**
closes #100

### CODE REVIEW

File: lib_comfyui/parallel_utils.py [hunk 1]
```diff
@@ -1,123 +1,292 @@
...
-        if self.producer_thread is None:
+        with self._lock:
+            self._memory_event.start()
+
+        self._lock.stop()
+
+    def send(self, value: object):
+        with self._lock:
+            self._filepath.unlink(missing_ok=True)
+            with RestoreTorchLoad():
```
Comment (ljleb) [inline lib_comfyui/parallel_utils.py:131]: unnecessary

File: lib_comfyui/parallel_utils.py [hunk 2]
```diff
@@ -1,123 +1,292 @@
...
+        self._producer_thread.join()
+        self._producer_thread = None
+        self._queue.stop()
 
     def is_running(self):
-        return not self.stop_event.is_set()
+        return self._producer_thread and self._producer_thread.is_running()
 
 
 class CallbackQueue:
```
Comment (ljleb) [inline lib_comfyui/parallel_utils.py:37]: not a queue anymore.

File: lib_comfyui/parallel_utils.py [hunk 3]
```diff
@@ -1,123 +1,292 @@
...
+        self._observer = None
+
+    def acquire(self, timeout=None):
+        while True:
+            if self._event.wait(timeout):
+                try:
+                    with open(self._filepath, 'x'):
+                        self._event.clear()
+                        return True
+                except FileExistsError:
```
Comment (ljleb) [inline lib_comfyui/parallel_utils.py:266]: add back comment or refactor the code

**Comment by ljleb** [comment]
parallel utils needs unit tests. Otherwise there is no way for me to asses code quality, it's too complex.

**Comment by ljleb** [comment]
I am running into this issue: https://github.com/python/cpython/issues/82300
I'll see if I can unregister the shared memory object

**Comment by ljleb** [comment]
Finally got `portalocker` to do 90% of the job. This only requires reading and writing to 4 files, as opposed to +16.
We could add an option to communicate using shared memory while keeping track of the state of the shared memory with the lock files. This is necessary for anyone hosting the default tempdir of their system on an SSD.

**Comment by ljleb** [comment]
What remains to be done:
- Don't unnecessarily run a workflow when it does not do anything. I think we can do this in a follow-up PR as it is not a priority
- [ ] Update the functional tests to use subprocesses instead of a single process.

**Comment by John-WL** [comment]
I'll wait until the conflicts are resolved to review this

**Comment by John-WL** [comment]
I would put all `ipc` and `ipc_*` files under the folder `ipc`

**Comment by ljleb** [comment]
Sure. Do I leave `ipc` like that, rename or also nest into `ipc` module

**Comment by John-WL** [comment]
> Sure. Do I leave `ipc` like that, rename or also nest into `ipc` module

you can leave nest it without renaming it, as it's sort of the root of the tree for these files

File: lib_comfyui/ipc_strategies.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+
+__all__ = [
+    'FileSystemIpcStrategy',
+    'SharedMemoryIpcStrategy',
+    'OsFriendlyIpcStrategy',
+]
+
+
+class FileSystemIpcStrategy:
```
Comment (John-WL) [inline lib_comfyui/ipc_strategies.py:16]: do you think this and `SharedMemoryIpcStrategy` would benefit from having a superclass?

Comment (ljleb) [reply lib_comfyui/ipc_strategies.py:16]: yeah

**Review by John-WL — APPROVED** [review]
LGTM

**Pull request merged**